### ISSUE TITLE
Operator overloading?

### ISSUE THREAD

**Issue by mramaiya** [ninia/jep#385]
So, I have a Java class that have the equivalent operator methods set, and I'm trying to then use the operators in Python. This works fine in python (if I wrote the class in python) and it works fine in Jython. If I call the __add__ method directly, that works as well, but for some reason, if I do a + b, for some reason, it's unable to execute the operator method.

```
package cff.python.interpreter;

public class Signal {

    private final int n;

    public Signal() {
        this. n = 0;
    }

    public Signal(int n) {
        this. n = n;
    }

    public Signal __add__(int i) {
        return new Signal(this.n + i);
    }

    @Override
    public String toString() {
        return "Signal:" + n;
    }
}

```

```
package cff.python.interpreter;

import jep.Interpreter;
import jep.JepConfig;
import jep.SharedInterpreter;

public class Main {

    public static void main(String[] args) {
        SharedInterpreter.setConfig(new JepConfig().redirectStdout(System.out));
        try (Interpreter interpreter = new SharedInterpreter()) {
            interpreter.exec("""
                    from cff.python.interpreter import Signal;
                    s = Signal(1);
                    print(dir(s))
                    print(s.__add__)
                    print(s.__add__(4))
                    print(s + 1)                     
                    """);
        }
    }

}
```

I'm wondering if I'm missing something simple here.

**Comment by bsteffensmeier** [comment]
Unfortunately, this is not currently working on jep. For operator overload to work in general python needs the method objects to exist at the time of type creation but jep does not add them until object creation. I think it would be possible to change jep to add the method objects to the type dict before type creation but it would be a significant change. It is something I would like to add to jep at some point but I have not yet found the time to do so. There is some slightly related discussion about overriding __repr__ on #339.

**Comment by mramaiya** [comment]
Got it. Thanks for the quick response. 

I looked through the code to see if I could implement it myself, but knowledge of C isn't strong enough to warrant it.

**Comment by bsteffensmeier** [comment]
I've rearranged the code so methods are added during type creation and added a test to verify that an `__add__` method in java enables operator overload in python. The code is available in #386, if there are no problems with the change then I am hoping to get this capability in the next jep release.

@mramaiya If you have the ability to test this and make sure it works as expected it would be appreciated.

**Comment by mramaiya** [comment]
Oh, man, that would be awesome. I really appreciate it.

Yeah, I’ll test it out. I’m guessing this would work for all the overloadable methods, right?

Milan

**Comment by ndjensen** [comment]
Closing as fixed in Jep 4.1.0.
